std.range: explain difference between lockstep and zip

greenify opened the pull request:
this is a follow-up to the discussions from the [forum](http://forum.dlang.org/post/qypadwjlzpktuqawbsza@forum.dlang.org)

1) I tried to add an explanation for the differences 
2) I removed the `foreach` example for `zip` - this is too confusing!
3) I had to add the remaining `zip` unittest in the documentation as otherwise the order with See_Also would be messed up :S

ntrel commented on std/range/package.d:
```diff
@@ -3515,8 +3515,26 @@ private alias lengthType(R) = typeof(R.init.length.init);
-   ranges in lockstep. For example, the following code sorts two arrays
-   in parallel:
+   ranges in lockstep.
+   For example, the following code sorts two arrays in parallel:
```
I think we could move this sentence to a comment at the top of the unit test, then the example doesn't need to be in the docs.

greenify replied:
agreed - just didn't know a nice way to fit it into the text. I will try again ;-)

ntrel commented on std/range/package.d:
```diff
@@ -3877,7 +3894,6 @@ pure unittest
     assert(b == [ "c", "b", "a" ]);
```
Perhaps `sort!greaterThan` would be easier to read, importing std.functional.

ntrel replied:
I suppose `t1, t2` might be clearer than `c, d` as they are tuples. Also I think UFCS style would be better: `zip(...).sort!(...)`.

greenify replied:
This is the already existing unittest example, but yeah it looks at lot nicer :)

greenify replied:
> Perhaps sort!greaterThan would be easier to read, importing std.functional.
> edit: sorry, ignore this comment, that wouldn't work.

It does work, because numerics are before strings.
However I think that the key point of this example is more to show how to sort array b according to a. That's why I swapped "b" and "c", so that this becomes more obvious.

ntrel commented on std/range/package.d:
```diff
@@ -3877,7 +3894,6 @@ pure unittest
     assert(b == [ "c", "b", "a" ]);
 }
 
-///
```
I don't think we should undocument this, I actually think it should come before the sort example.

ntrel replied:
Similar to the sort example, I suggest we rename _e_ to _tup_ as it's a tuple.

greenify commented on std/range/package.d:
```diff
@@ -3515,8 +3515,26 @@ private alias lengthType(R) = typeof(R.init.length.init);
+   See_Also: $(LREF lockstep)
+
+      $(LREF lockstep) is similar to $(D zip), but it bundles its values
+      and returns a range. $(LREF lockstep) is useful for
```
I updated the comment strings, please have a look again. This time I hope I didn't mess up the description ;-)

greenify commented:
Please have a look at example 2. There are three ways that we could recommend - they all work.
Which should we recommend? I think all three are valid and could be useful.

As we now have an entire PR about it. Let me clarify - we have `locksteep`  __only__ because it is faster for loops?

```
    import std.algorithm : equal, map;
    import std.conv: to;

    int[] a = [ 1, 2, 3 ];
    string[] b = [ "a", "b", "c" ];

    size_t idx = 0;
    foreach (tup; zip(a, b))
    {
        assert(tup[0] == a[idx]);
        assert(tup[1] == b[idx]);
        ++idx;
    }

    idx = 0;
    foreach (t1, t2; zip(a, b))
    {
        assert(t1 == a[idx]);
        assert(t2 == b[idx]);
        ++idx;
    }

    auto result = zip(a, b).map!(tup => tup[0].to!string ~ tup[1]);
    assert(result.equal(["1a", "2b", "3c"]));
```

Moreover I added another nice example for `zip` - it's taken from my D Functional Garden and I think it's a nice way to show how zip works:

```
// pairwise sum
auto arr = [0, 1, 2];
assert(arr.zip(arr.save.dropOne).map!"a[0] + a[1]".equal([1, 3]));
```

ntrel commented:
LGTM :+1:

greenify commented:
@ntrel - thanks for your helpful feedback and review.

The examples now look a lot better:

http://dtest.thecybershadow.net/artifact/website-23160046fb7ccd78e54d467d28bdb87a4c8afd25-53cf8893d4f37218a84d349547178895/web/phobos-prerelease/std_range.html#.zip

side note: the formatting of the ddocs output could be improved a bit (e.g. no duplicates "Examples")- see [issue 15769](https://issues.dlang.org/show_bug.cgi?id=15769)

JakobOvrum commented:
The actual reason `lockstep` isn't deprecated is because it allows reference access to elements:
```d
auto left = new int[](3);
auto right = new int[](3);
foreach(ref lhs, ref rhs; lockstep(left, right))
{
	lhs = 1;
	rhs = 2;
}
assert(left == [1, 1, 1]);
assert(right == [2, 2, 2]);
```
External iteration ala `zip` is superior in every other respect.

greenify commented:
> The actual reason lockstep isn't deprecated is because it allows reference access to elements:

Okay then I will update the docs to point this out even more clearly :)

Btw having idexing in foreachs `opCalls` doesn't work with zip yet, but this might be easy to add?

JakobOvrum commented:
What do you mean? If you mean iterating with an index, std.range.enumerate is designed to work well with zip.

greenify commented:
> What do you mean? If you mean iterating with an index, std.range.enumerate is designed to work well with zip.

Yes I know, but I was more referring that `lockstep` support the "direct" style via overloads::

```
    import std.stdio;
    import std.range;
    auto a = [1, 2, 3];
    auto b = ["a", "b", "c"];
    foreach(i, t1, t2; zip(a, b).enumerate)
    {
        writeln(i, t1, t2);
    }

    foreach(i, t1, t2; lockstep(a, b))
    {
        writeln(i, t1, t2);
    }
```

greenify commented:
> Okay then I will update the docs to point this out even more clearly :)

Updates:

- added a similar example example with `ref` to locksteps documentation
- added info about access with references to the explanation about differences (thanks @JakobOvrum)
- replaced `values` with `elements`

ntrel commented on std/range/package.d:
```diff
@@ -3869,27 +3874,55 @@ auto zip(Ranges...)(Ranges ranges)
+
+    idx = 0;
+    // it supports automatic tuple unpacking too
+    foreach (t1, t2; zip(a, b))
```
Here I suggest we use `e1, e2` as they're not tuples.

greenify replied:
done - I also renamed the loop parameter for `lockstep`

ntrel commented:
@greenify :
> I was more referring that lockstep support[s] the "direct" style via overloads:
> foreach(i, t1, t2; lockstep(a, b))

I think that's due to lockStep's opApply. I'm not sure that tuple unpacking for zip would work correctly if there was an index element but no index variable. 

BTW foreach tuple unpacking isn't documented AFAICT:
https://issues.dlang.org/show_bug.cgi?id=7361

greenify commented:
> So I think it's better to not document this anti-feature.

That's the wrong way in my opinion, if it is slow due to the tuple expansion, we should either

- document _that_ 
- disable it

and not hope that people might not try it.

ntrel commented:
> added a similar example example with ref to locksteps documentation

There already was an example with `ref` ;-) Please see my pull which tweaks some things and also avoids docstring examples: https://github.com/greenify/phobos/pull/1

greenify commented:
> Please see my pull which tweaks some things and also avoids docstring examples: greenify#1

Thanks @ntrel - merged with this branch :)

ntrel commented:
Simplify zip examples more: greenify#1

greenify commented:
Should I squash the commits?

ntrel commented:
Sure, go ahead.

greenify commented:
squashed and rebased to master.

@ntrel - it's a shame git doesn't allow joint commit messages :/

greenify commented:
ping @ reviewers - @ntrel and I have worked on this minor update to the documentation which yields a better understanding of the difference between `lockstep` and `zip` to the reader and adds better examples. Imho this trivial PR is ready now - any other thoughts or blocking issues?

ntrel commented:
@greenify I split out the obvious example changes into a separate PR but kept your authorship for the commit, hope this is OK: #4078. I left out anything which might need more reviewing.

greenify commented:
> hope this is OK

Absolutely - you could/can change it back to your name. I don't mind either. 

Thanks for pushing this @ntrel, but I guess the main reason the inactivity is due to the weekend?

JakobOvrum commented:
Auto-merge toggled on
